### ISSUE TITLE
Pull request reviews

### ISSUE THREAD

**Issue by johanneskoester** [bioconda/bioconda-recipes#2327]
Dear @bioconda/core,
shall we enable [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/)? The consequence would be that PRs cannot be merged before one other dev has added an approval without further requests.

This would be great for quality control, but maybe slow down some PRs which only update the version of a tool. Nevertheless, I am in favor of trying this out. What do you think?

**Comment by bgruening** [comment]
As discussed at GCC I'm in favor of this + working on more strict travis quality control!

**Comment by chapmanb** [comment]
Sorry for always being negative recently, but I'm -1 on this. Being able to quickly merge fixes/updates is one of the things I especially like about the bioconda community process. I'm against blocking people by forcing every update to get a review, but for encouraging reviews when contributors would like them. Since testing catches lots of issues and it's pretty easy to provide a fix for package issues by pushing a new version, I'd vote for keeping the current laissez-faire approach which encourages and enables contributors. What type of submissions are we trying to catch/avoid by adding mandatory reviews?

**Comment by johanneskoester** [comment]
I see your point, Brad. I think mandatory reviews would mostly avoid problems with missing pinning or missing dependencies (e.g. gcc). But alternatively we could enforce that via linting as well. Another possibility would be to try to enable mandatory reviews only for new recipes (e.g. via adding an automatic review via the API in case of update-only PRs).

On the other hand, mandatory reviews for everything could also actually decrease the quality of reviews, because the workload will increase...

**Comment by bgruening** [comment]
@chapmanb the largest concern about bioconda I frequently hear is that this channel is not secure. Everyone can basically change everything. It is easy to trick travis and inject some code, include a new numpy version that breaks everything else. And because most of the versions are not pinned this has the potential to break a lot of packages.

@s-will thought already about signing packages to distinguish between trusted and non-trusted packages, but this is complicated - yet support in conda is already there.

The next obvious thing is to improve the conda tests, but every test you can trick I assume. 
I see the most harmful potential in new packages - so I think at least these should be checked? No?

**Comment by kyleabeauchamp** [comment]
I also worry about the security concern that @bgruening brings up.

**Comment by tomkinsc** [comment]
Security of the channel concerns me as well. Could we have a mechanism to require code reviews until a user has a certain number of accepted PRs? We could add them to a particular organizational group after so many successful merges, under the assumption they'd be trusted at that point. Routine contributors would be able to make quick fixes by approving their own PRs, but new contributors would need to prove themselves first. It wouldn't be an onerous gate, but it would prevent the (lovely) free-for-all state we have now. So far the current system has worked out, but as bioconda becomes more successful it will also become a bigger target for bad actors.

**Comment by johanneskoester** [comment]
We could stop adding people immediately. Instead, we could only provide invitations after x successful PRs.

**Comment by croth1** [comment]
The github infrastructure should work very well with only a small number of trusted core members that review and merge pull requests. This will not only improve security but will also help to ensure quality standards of our recipes in the future.
I think this model would be much better, provided the core team is large enough so that waiting periods for reviews are reasonably short.

Users that contribute a lot and rely on quickly introducing updates/changes as @chapmanb mentioned, could simply become trusted core members and then enjoy the right to pull in their changes immediately.

**Comment by hainm** [comment]
I am +1 for pull request review and completely agree with @bgruening about

> Everyone can basically change everything. It is easy to trick travis and inject some code, include a new numpy version that breaks everything else. And because most of the versions are not pinned this has the potential to break a lot of packages.

I agree with @tomkinsc about this. And may be the core developer here can also review new user's profile to either prove or defer (until they prove themselves).

> Routine contributors would be able to make quick fixes by approving their own PRs, but new contributors would need to prove themselves first.

**Comment by chapmanb** [comment]
Thanks all for the comments and the great discussion. I appreciate the potential concerns about security and other issues, but still favor a lightweight process to push updates and recipes. Contributors will try to do the right thing and we can handle bad actors on a single case basis if they occur. The vibrant bioconda community exists because of the way we've removed barriers to contributing, and this has resulted in a large set of contributors that push fixes and improvements to recipes. Adding rules and process unfortunately will discourage this.

Having a vetting period before giving folks commit access and submitting a new PR as an introduction is a good idea. We could also encourage recipe reviews and I think most people would use this until they feel comfortable contributing directly. I'd much prefer adding documentation on doing the right thing, rather than introducing more barriers and work for contributors.

**Comment by johanneskoester** [comment]
That's a reasonable argument. I must say that I am unsure what approach is better. Both have advantages and disadvantages. In a perfect world, we would have an easy way to enable forced reviews only for new recipes...

Maybe we make a two-step process out of this? I guess anybody is ok with inviting only after the first successful PR? That means, people posting in the join thread would post a link to the successful PR, and we add them.

Once that appears to work and is established, we can re-evaluate if we need even more security.

**Comment by s-will** [comment]
I appreciate this discussion very much as well. Also, I share the concerns about making contributions to bioconda more complicated. However, IMO security is extremely important for the success of the bioconda project. Some time ago, I wrote down some ideas, how security could be fundamentally improved while keeping updates and pushing of new recipes as lightweight as possible. I want to share these ideas here, because it fits part of the discussion well and maybe adds  a new perspective.

My ideas about protecting Bioconda packages against manipulation and preventing malicious packages on Bioconda are based on signing recipes and packages. The core idea is to build up a system for secure packages (signed by trusted contributors) hat works in parallel to the existing process. Administrators could then choose to install only trusted packages on critical systems etc.

Here is what I wrote in the hope it is helpful; I am curious about your comments. 


Signing bioconda packages
=========================
[Below, I integrated some comments thanks to @bgruening.]

Motivation
----------
Since Bioconda is an open community project, a growing number of
contributors have rights to create and update recipes. This lets
Bioconda grow fast but also makes it particularly vulnerable.  Thus,
packages could easily be manipulated e.g. for attacking computer
systems that install and update software from conda.  This endangers
the project, since successful attacks would likely have severe
impacts.

Cryptographic signatures can limit such risks significantly and allow
to build a trust system in lines with the open project structure.

Currently most major Linux distributions make use of packages
signatures to verify bin-packages in their repositiories. Protecting packages 
by signatures thus became state-of-the-art and will be expected by users. 

Having signed packages is also discussed by conda-forge e.g.
https://conda-forge.hackpad.com/conda-forge-meetings-2YkV96cvxPG
https://github.com/conda-forge/conda-forge.github.io/issues/129

Package signing and verification is supported by Conda
http://conda.pydata.org/docs/signed-packages.html


Goals
-----
  - Implement a 'community-based' signing mechanism for 'trusted' Bioconda packages
  - 'Trusted' packages on Bioconda receive verifyable cryptographic signatures
  - Users can verify that 'trusted' packages from bioconda are
      - not compromised by third parties (e.g. man-in-the-middle)
      - and can be trusted / are non-malicious according to trusted contributors
  - Trusted and untrusted packages can coexist on bioconda
  - Efforts for establishing trust in packages are scalably distributed to the community
  - The mechanism does not compromise  the project's openness and ease of contributing

Existing functionality in conda
-------------------------------
Conda implements a signing/verification functionality with private/public keys

- key generation

     conda sign --keygen

- signing

     conda sign file

- verfiying

     conda sign --verify file

This enables signatures for single recipe files as well as for conda packages
    (tar.bz2 files)

Letting 'conda install' install only signed packages seems to be implemented at least as preview:
http://conda.pydata.org/docs/signed-packages.html


Proposed working mechanism
--------------------------
- Travis automatically signs bin packages (with bioconda key),
    - if (new) recipe is signed and trusted
    - if (existing) recipe gets signed and trusted
  NOTE: no need to trust run-time-only dependencies in this step,
        this avoids recursive checking of deps for signing a single package;
        build dependencies have to be checked recursively

- The signatures are provided as files *.tar.bz2.sig along with the
     conda package *.tar.bz2 in the bioconda channel [@bgruening raised the question: is
     this supported in the bioconda channel hosted by anaconda.org?]

- Contributors sign recipes that they create (or alter or have verified)
    - all files in the recipe are signed separately (w/ conda sign)
          this comprises meta/build/patches/...?, basically everything but
            - source code (which is secured by checksum) 
            - dependencies
    
- Maintain a list of trusted contributors in bioconda's git repository (github/bioconda/keys):
    checkin public keys of contributors to bioconda (use github names)
    meta-trusted contributors can sign public keys to make them trustworthy

- Use the signatures
  - Extend conda install such that it installs only signed packages
     - this includes verifying all run-time dependencies
     - this functionality seems to be implemented (see Step 5 of http://conda.pydata.org/docs/signed-packages.html)
  - Provide a simple mechanism to verify the already installed packages (including dependencies)


Optional ideas to extend the system
-----------------------------------
    - Recipe signing could be refined by allowing several
      signatures per file by different contributors; however, multiple
      signatures are not necessary for a working system and thus can
      be postponed or omitted.

      - would allow trusted contributors to sign recipes of 
        not-yet-trusted contributors ...
      - would allow web of trust ideas
      - optionally this would be extensible by trust levels
      - this is currently NOT well supported by 'conda sign', since conda sign does
        neither support multiple signatures in one file nor signature files with names other
        than filename.sig
    - List of trusted contributors could be annotated with individual trust levels




Additional issues
-----------------
- Many conda packages from other channels are currently unsigned, which will not necessarily change very soon.
  This would currently prevent the secure installation of (trusted) packages, which depend on them.
  Some ideas to handle this should be discussed:
     - Rely on (or try to convince) other channels to host signed, trusted packages as well
     - For build-dependencies, generally trust certain channels like conda-forge;
       This seems justifiable, if such channels are inherently less vulnerable than bioconda 
     - Trust entire channels like conda-forge for run-time dependencies (this can be decided by the user!)
     - Mirror trusted third-party recipes/packages and sign them. 
       This seems easy to combine with other ideas; could be useful for selected packages but adds
       load on the bioconda community.
     - Host signatures for non-bioconda conda packages [probably not
       the best idea, since this would not be very useful without
       changing conda's signature verification]
        
- Can we easily implement useful KEY REVOCATION, i.e. /untrusting/
  contributors? This could be done by removing the meta-trust
  signature from the contributor's pubkey in the bioconda github
  repository. In turn, Travis could delete the signature; maybe
  generating a new unsigned release of the newly untrusted binary
  packages. However, this would not be perfect, since signatures and
  tarballs were already published.  To increase security, conda could
  maintain a black list of untrusted packages.

- Policies are required to define the meaning of a recipe signature
      - For example: the signee of a meta file has checked that the source code
         stems from the site of the authors
      - Contributors who sign their packages implicitely sign the policy 
         (of course: with disclaimer of responsibility for damages etc; there have to
         be limits)

- All critical components of this system (e.g. bioconda travis ci) have to be
  protected against untrusted contributors

**Comment by johanneskoester** [comment]
@s-will indeed, signing packages would be a good solution, and your concept sounds reasonable. I fully support this, although it sounds like conda (or anaconda) is not quite there yet, right?

**Comment by s-will** [comment]
Hi @johanneskoester, you are right, conda is not there yet, but it looks good that this will be completely supported soon. 
I just found this tweet from Kale Franz, where he announces such functionality for end of 2016/early 2017: https://twitter.com/kalefranz/status/741094414085435392 
(TUF-The Update Framework is a framework to facilitate making software update systems secure).
I found other 'hints' in discussions of  conda-forge (https://conda-forge.hackpad.com/conda-forge-meetings-2YkV96cvxPG) or in https://www.reddit.com/r/Python/comments/4766ck/is_anaconda_secure/

Admittedly, it would still be great to hear from the conda developers directly what's actually going on. Do we have good connections and can find out more?
Still, to me it seems the right time to start this discussion and prepare for the coming "secure conda"; if these possibilities are finally there, Bioconda should probably have ideas ready, how package signatures can be married with the open contribution model.

**Comment by jerowe** [comment]
As a contributor I wouldn't mind if my recipes had to be reviewed before merging. Generally I ask for input anyways. The only recipes I've merged without asking for input have been things like perl/python packages, or just a quick version update of an existing recipe, but I wouldn't mind having a review process for any case. I'm not sure when @bgruening sleeps, because he nearly always reviews my packages right away. ;-)

I generally think its a good idea. Most other projects out in the wild don't allow absolutely everyone to approve PRs and they all get along just fine.

**Comment by bgruening** [comment]
@s-will have a look at: https://conda-forge.hackpad.com
Add your topic and join the next meeting :)

**Comment by bgruening** [comment]
@johanneskoester do we agree to invite people as soon as they have contributed a few packages? This would mean closing #1 and writing a statement into the readme about this new mechanism.

**Comment by johanneskoester** [comment]
We can do that, if the rest of @bioconda/core is fine with the change. What number of PRs should we require?
I would not close the issue. Rather I would change the text that the user requesting invitation shall quote n successful Bioconda PRs. This way, we don't need to track this information ourselves.

**Comment by daler** [comment]
Not sure if PR count is the right metric, since successfully merging a conda skeleton recipe might give false confidence. On the other hand it's not like we have a 'difficulty of recipe' score, so PR count is probably the best we have.

1 PR is pretty close to what we have now because there are a quite a few people that join and then submit one PR. So we would just be switching the order, which would hopefully be not too disruptive. One PR, asking for review and optionally asking to join, sounds good to me.

**Comment by epruesse** [comment]
closing as stale

Not that the point made here is irrelevant. Should be split into separate issues on bioconda-utils with label 'security' if anyone wants to pick this up.
